File: Hangman Equivalency/Hangman.cpp
#include <iostream>
#include <string>
#include <vector>
#include <iomanip>
#include <algorithm>
#include <regex>
#include <thread>
#include <chrono>

void clearConsole(void)
{
	 std::system("cls"); //I was told this was bad for security reasons. Oh well.
}

int main()
{
	 bool running = true;
	 while (running)
	 {
			clearConsole();
			bool gaming = true;
			bool validInput = false;
			std::string fullWord;
			std::string guessWord;
			std::vector<char> guessedLetters;
			int livesLeft = 6;
			std::cout << "Welcome to hangman! Enter a word to start: (Can contain only letters and spaces.) \n";
			while (!validInput)
			{
				 fullWord.clear();
				 std::getline(std::cin, fullWord);
				 // Cleans the input.
				 // Split on spaces, remove entry entries, trim whitespace of words, lowercaseify.
				 std::transform(fullWord.begin(), fullWord.end(), fullWord.begin(), [](unsigned char c) { return std::tolower(c); });
				 std::vector<std::string> temp;
				 for (size_t ind = 0; ind < fullWord.length(); )
				 {
						ind = fullWord.find_first_not_of(' ', ind);
						size_t end = fullWord.find(' ', ind);
						end = (end == ((size_t)-1)) ? fullWord.length() : end; //no -1 on len to correct for error on the last iteration
						size_t len = (end - ind);
						temp.push_back(fullWord.substr(ind, len));
						ind = end;
				 }
				 fullWord.clear();
				 for (auto& e : temp)
				 {
						fullWord += (e + " ");
				 }
				 if (fullWord.length() > 0)
				 {
						fullWord.erase(fullWord.length() - 1);
				 }
				 if (std::regex_match(fullWord.c_str(), std::regex("^[a-zA-Z ]+$")))
				 {
						validInput = true;
				 }
				 else
				 {
						std::cout << "Sorry, that input isn't valid. You can only choose a word or phrase that contains letters and spaces.\n";
						std::cout << "Enter a word to start: (Can contain only letters and spaces.) \n";
				 }
			}
			// We now have a valid input.
			guessWord.resize(fullWord.length(), ' ');
			std::transform(fullWord.begin(), fullWord.end(), guessWord.begin(), [](unsigned char c) { return ((c == ' ') ? ' ' : '_'); });
			clearConsole();
			//print status bar:
			std::cout << "Lives Left: " << std::to_string(livesLeft) << ".   Letters guessed: ";
			for (size_t i = 0; i < guessedLetters.size(); i++)
			{
				 std::cout << guessedLetters[i];
				 if (i + 1 != guessedLetters.size())
				 {
						std::cout << ", ";
				 }
			}
			std::cout << "\n\n";
			//print word:
			std::cout << guessWord << "\n";
			while (gaming)
			{
				 //print prompt:
				 bool validGuess = false;
				 char guess = '\0';
				 while (!validGuess)
				 {
						std::cout << "\nGuess a letter: \n";
						std::string temp;
						std::getline(std::cin, temp);
						guess = ((temp.length() > 0) ? temp[temp.find_first_not_of(' ')] : guess);
						guess = std::tolower(guess);
						if (guess == '\0')
						{
							 std::cout << "Make sure to type a character. The first character entered will be your guess.\n";
						}
						else if (std::count(guessedLetters.begin(), guessedLetters.end(), guess) > 0)
						{
							 std::cout << "You already guessed that letter!\n";
						}
						else if (!std::isalpha(guess))
						{
							 std::cout << "You can only guess English letters!\n";
						}
						else
						{
							 validGuess = true;
						}
				 }
				 guessedLetters.push_back(guess);
				 //evaluate:
				 bool didFindLetter = false;
				 for (size_t i = 0; i < fullWord.length(); i++)
				 {
						if (std::tolower(fullWord[i]) == guess)
						{
							 didFindLetter = true;
							 if (i == 0 || fullWord[i - 1] == ' ')
							 {
									guessWord[i] = std::toupper(guess);
							 }
							 else
							 {
									guessWord[i] = std::tolower(guess);
							 }
						}
				 }
				 if (!didFindLetter)
				 {
						livesLeft--;
				 }
				 //Clear for fresh start.
				 clearConsole();
				 //print status bar:
				 std::cout << "Lives Left: " << std::to_string(livesLeft) << ".   Letters guessed: ";
				 for (size_t i = 0; i < guessedLetters.size(); i++)
				 {
						std::cout << guessedLetters[i];
						if (i + 1 != guessedLetters.size())
						{
							 std::cout << ", ";
						}
				 }
				 std::cout << "\n\n";
				 //print word:
				 std::cout << guessWord << "\n";
				 //check for win/lose.
				 if (livesLeft == 0 || std::count(guessWord.begin(), guessWord.end(), '_') == 0)
				 {
						if (livesLeft == 0)
						{
							 //You lose.
							 std::cout << "Uh oh! Out of lives, you lose. The word/phrase was: \"" << fullWord << "\" Play Again? (Y/N)\n";
						}
						else
						{
							 //You win!
							 std::cout << "Congratulations! You win! Play Again? (Y/N)\n";
						}
						std::string input;
						validInput = false;
						while (!validInput)
						{
							 std::getline(std::cin, input);
							 if (std::regex_match(input.c_str(), std::regex("^[Y|yE|eS|s]|[Y|y]")) || std::regex_match(input.c_str(), std::regex("^[N|nO|o]|[N|n]|[Q|qU|uI|iT|t]")))
							 {
									validInput = true;
							 }
							 else
							 {
									std::cout << "Hey! That isn't a valid input, say \"Y\" or \"N\".\n";
							 }
						}
						if (std::regex_match(input.c_str(), std::regex("^[Y|yE|eS|s]|[Y|y]")))
						{
							 livesLeft = 6;
							 guessedLetters.clear();
							 guessWord.clear();
							 fullWord.clear();
							 gaming = false;
							 running = true;
						}
						else if (std::regex_match(input.c_str(), std::regex("^[N|nO|o]|[N|n]|[Q|qU|uI|iT|t]")))
						{
							 std::cout << "Thanks for playing!\n";
							 std::this_thread::sleep_for(std::chrono::seconds(2));
							 gaming = false;
							 running = false;
						}
				 }
			}
	 }
}

File: Hangman Equivalency/Hangman.vb
Imports System
Imports System.Linq
Imports System.Text.RegularExpressions
Imports System.Text
Imports System.Collections
Imports System.Collections.Generic
Imports System.Threading

Module Hangman
   Private ReadOnly alphaAndSpaceReg As Regex = New Regex("^[a-zA-Z ]+$")
   Private ReadOnly alphaCharReg As Regex = New Regex("[a-zA-Z]")
   Private ReadOnly affirmativeReg As Regex = New Regex("^[Y|yE|eS|s]|[Y|y]")
   Private ReadOnly negativeReg As Regex = New Regex("^[N|nO|o]|[N|n]|[Q|qU|uI|iT|t]")
   Sub Main(args As String())
      Dim running As Boolean = True
      While running
         Console.Clear()
         Dim gaming As Boolean = True
         Dim validInput As Boolean = False
         Dim fullWord As String = String.Empty
         Dim guessWord As StringBuilder = New StringBuilder(String.Empty)
         Dim guessedLetters As List(Of Char) = New List(Of Char)
         Dim livesLeft As Int32 = 6
         Console.WriteLine("Welcome to hangman! Enter a word to start: (Can contain only letters and spaces.) ")
         While Not validInput
            fullWord = String.Join(" ", Console.ReadLine().Trim().Split(" ", StringSplitOptions.RemoveEmptyEntries)).ToLower()
            If alphaAndSpaceReg.IsMatch(fullWord) Then
               validInput = True
            Else
               Console.WriteLine("Sorry, that input isn't valid. You can only choose a word or phrase that contains letters and spaces.")
               Console.WriteLine("Enter a word to start: (Can contain only letters and spaces.) ")
            End If
         End While
         ' We Now have a valid input.
         guessWord.Append(alphaCharReg.Replace(fullWord, "_"))
         Console.Clear()
         ' print status bar:
         Console.WriteLine($"Lives Left: {livesLeft}.   Letters guessed: {String.Join(", ", guessedLetters)}{Environment.NewLine}")
         ' print word
         Console.WriteLine(guessWord.ToString())
         While gaming
            ' print prompt:
            Dim validGuess As Boolean = False
            Dim guess As Char = Char.MinValue
            While Not validGuess
               Console.WriteLine($"{Environment.NewLine}Guess a letter: ")
               guess = Console.ReadLine().ToLower().Trim().FirstOrDefault()
               If guess = Char.MinValue Then
                  Console.WriteLine("Make sure to type a character. The first character entered will be your guess.")
               ElseIf guessedLetters.Contains(guess) Then
                  Console.WriteLine("You already guessed that letter!")
               ElseIf Not alphaCharReg.IsMatch(guess.ToString()) Then
                  Console.WriteLine("You can only guess English letters!")
               Else
                  validGuess = True
               End If
            End While
            guessedLetters.Add(guess)
            ' evaluate:
            Dim didFindLetter As Boolean = False
            For index = 0 To fullWord.Length - 1
               If fullWord.ToLower()(index) = guess Then
                  didFindLetter = True
                  guessWord.Remove(index, 1)
                  If index = 0 Then
                     guessWord.Insert(index, guess.ToString().ToUpper())
                  Else
                     If fullWord(index - 1) = " " Then ' can't add this as an or to the above if line, vb seems to not break if statement evaluations if the first half of an or is true.
                        guessWord.Insert(index, guess.ToString().ToUpper())
                     Else
                        guessWord.Insert(index, guess.ToString().ToLower())
                     End If
                  End If
               End If
            Next index
            If Not didFindLetter Then
               livesLeft -= 1
            End If
            ' Clear for fresh start.
            Console.Clear()
            ' print status bar:
            Console.WriteLine($"Lives Left: {livesLeft}.   Letters guessed: {String.Join(", ", guessedLetters)}{Environment.NewLine}")
            ' print word:
            Console.WriteLine(guessWord.ToString())
            ' check for win/lose.
            If livesLeft = 0 Or (Not guessWord.ToString().Contains("_")) Then
               If livesLeft = 0 Then
                  ' You lose.
                  Console.WriteLine($"Uh oh! Out of lives, you lose. The word/phrase was: ""{fullWord}"" Play Again? (Y/N)")
               Else
                  ' You win!
                  Console.WriteLine("Congratulations! You win! Play Again? (Y/N)")
               End If
               Dim input As String = String.Empty
               validInput = False
               While Not validInput
                  input = Console.ReadLine()
                  If affirmativeReg.IsMatch(input) Or negativeReg.IsMatch(input) Then
                     validInput = True
                  Else
                     Console.WriteLine("Hey! That isn't a valid input, say ""Y"" or ""N"".")
                  End If
               End While
               If affirmativeReg.IsMatch(input) Then
                  livesLeft = 6
                  guessedLetters.Clear()
                  guessWord.Clear()
                  fullWord = String.Empty
                  gaming = False
                  running = True
               ElseIf negativeReg.IsMatch(input) Then
                  Console.WriteLine("Thanks for playing!")
                  Thread.Sleep(2000)
                  gaming = False
                  running = False
               End If
            End If
         End While
      End While
   End Sub
End Module